### DISCUSSION TITLE
Web 项目热加载

### DISCUSSION THREAD

**Discussion by ghost** [coder/code-server#6105] [Ideas] [answered]
How to configure 'vue. config. js' without setting it

Instead, use environment variables or Code Server command parameters to set the URL of the Web Socket?

Similar to Github Codespace, there is no need to modify the project itself

Or where can I find detailed configuration documents for Code Server? 🥺🥺🥺

thank you 🙏

**Comment by code-asher** [comment]
This is something that works in Codespaces but not code-server?  Could you provide an example or steps to reproduce?

**Reply by ghost** [reply]
First, I created a Vue project for testing and started it

The first visit will prompt

![7a698331bf98b57125f036d95bf9d4b3](https://user-images.githubusercontent.com/66768066/230086227-d01f6045-186a-4e23-a40d-3d76c9873bb1.png)
![9f2f97dfbd4d9b7e292a06d055ffa383](https://user-images.githubusercontent.com/66768066/230085655-aad4f6cb-b4d1-40a4-be94-284b9120bc8e.png)
So I need to modify my vue.config.js, add allowed domain name information, and then restart access
<img width="1097" alt="image" src="https://user-images.githubusercontent.com/66768066/230086834-3f09685b-8dad-4d46-ba4b-5f0a53839e9a.png">
Then I can successfully access my Vue project, but there will be new issues where the webstock connection address is localhost
![cc997e1aa1d8b3a5fb9a6d821a6d8c92](https://user-images.githubusercontent.com/66768066/230087457-570efc7b-fd85-4716-9afc-938e9ec05041.png)
So I also need to configure the websocket address of the Vue project and restart it to use it like the desktop version
![d6f1c30ba614f99fcfe5cb23c2d22155](https://user-images.githubusercontent.com/66768066/230088060-985c256a-90a1-4779-a05c-f390b43f1278.png)
![18658c37975685e0e34fb0e401b4236b](https://user-images.githubusercontent.com/66768066/230088094-b49ceb13-1eaa-470f-addf-33bc5bf04f99.png)
But in Github Codespaces, these additional configurations are not required and can be used directly

**Reply by code-asher** [reply]
Thank you for the steps!

How are you sending `8848.jojo.codespace.com` to code-server?  Looks like Vue wants the host header but it is missing.  If you are using something like NGINX you might need `proxy_set_header Host $host;`.

**Reply by ghost** [reply]
Yes, my nginx has already set up a proxy_ set_ header Host $host;
So how can you use it like a desktop version without needing to configure it in the Vue project?
Are you ming bai?

**Reply by code-asher** [reply]
Interesting, the `Invalid Host header` would imply there is some issue there.  Maybe check the browser network tab and make sure the request is setting the right host header?

I have not tried behind a reverse proxy but I tried running it on localhost and it worked for me.  

Here is what I did:

```
$ npm install -g @vue/cli
$ vue create my-project # I selected the default options
$ code-server --port 8081 --proxy-domain *.localhost &
$ cd my-project
$ yarn serve
```

Then I went to http://8080.localhost:8081 and the app worked as expected in both Chromium and Firefox.

**Reply by code-asher** [reply]
Ahhhhh wait yeah if I use a non-localhost host I get the same error.  Let me see if I can figure out why.

**Reply by code-asher** [reply]
Actually Codespaces is broken too, at least for me.  It does not show that white screen with "Invalid Host header" but the websocket does disconnect with `{"type":"error","data":"Invalid Host/Origin header"}`.

![ws](https://user-images.githubusercontent.com/45609798/231251088-6fa861bb-208f-4a6c-9987-8fc9bc987cb4.png)

**Reply by code-asher** [reply]
I echoed the headers Codespaces is sending to the app:

```
--> GET / HTTP/1.1
--> Host: localhost
```

It looks like Codespaces is dropping the host which is why the root page loads (this could technically be considered a bug in `webpack-dev-server` since they are not checking `X-Forwarded-Host` or `Forwarded`).  But then the web socket has an origin check which fails.  If Codespaces did not drop the host they would have the same problem we have.

https://github.com/webpack/webpack-dev-server/blob/abbcb448b72b2d11e07b656c8018c146d1688a91/lib/Server.js#L1967-L1977

https://github.com/webpack/webpack-dev-server/blob/abbcb448b72b2d11e07b656c8018c146d1688a91/lib/Server.js#L2553-L2558

So I think with Vue/webpack-dev-server it is expected you are supposed to configure the allowed hosts when behind a reverse proxy.

**Reply by code-asher** [reply, marked as answer]
While that explains why we need `allowedHosts` it does not explain why we need `webSocketURL`.  I poked at it a little but I still have no idea why it works in Codespaces without extra configuration.  For now I think we just have to set it.

That leaves us with this config:

```js
module.exports = defineConfig({
  transpileDependencies: true,
  devServer: {
    allowedHosts: "all",
    client: {
      webSocketURL: "auto://0.0.0.0:0/ws",
    },
  },
})
```
